mlir/lib/Dialect/Tosa/Transforms/TosaNarrowTypes.cpp

```diff
@@ -0,0 +1,558 @@
+  typeConverter.addConversion([&typeConverter](RankedTensorType type) -> Type {
+    Type elementType = type.getElementType();
+    if (!isSourceElement<Kind>(elementType))
+      return type;
+    Type converted = typeConverter.convertType(elementType);
+    if (!converted || converted == elementType)
+      return type;
+    return RankedTensorType::get(type.getShape(), converted,
+                                 type.getEncoding());
+  });
+  typeConverter.addConversion(
+      [&typeConverter](UnrankedTensorType type) -> Type {
+        Type elementType = type.getElementType();
+        if (!isSourceElement<Kind>(elementType))
+          return type;
+        Type converted = typeConverter.convertType(elementType);
+        if (!converted || converted == elementType)
+          return type;
+        return UnrankedTensorType::get(converted);
+      });
+
+  const auto materializeCast = [](OpBuilder &builder, Type resultType,
+                                  ValueRange inputs, Location loc) -> Value {
+    if (inputs.size() != 1)
+      return Value();
+    return tosa::CastOp::create(builder, loc, resultType, inputs.front());
+  };
+  typeConverter.addSourceMaterialization(materializeCast);
+  typeConverter.addTargetMaterialization(materializeCast);
+
+  if constexpr (Kind == TosaNarrowKind::Int64ToInt32) {
+    typeConverter.addTypeAttributeConversion(
+        [](IntegerType type, IntegerAttr attribute) -> Attribute {
+          Attribute convertedAttr;
+          if (tryConvertScalarAttribute<Kind>(attribute, convertedAttr))
+            return convertedAttr;
+          return attribute;
+        });
+    typeConverter.addTypeAttributeConversion([&typeConverter](
+                                                 ShapedType type,
+                                                 DenseIntElementsAttr attr)
+                                                 -> Attribute {
+      return convertDenseElementsAttr<Kind, DenseIntElementsAttr, IntegerType>(
+          type, attr, typeConverter,
+          [](IntegerType newElementType, const APInt &value) {
+            return value.truncSSat(newElementType.getWidth());
+          });
+    });
+  }
+
+  if constexpr (Kind == TosaNarrowKind::Float64ToFloat32) {
+    typeConverter.addTypeAttributeConversion(
+        [](FloatType type, FloatAttr attribute) -> Attribute {
+          Attribute convertedAttr;
+          if (tryConvertScalarAttribute<Kind>(attribute, convertedAttr))
+            return convertedAttr;
+          return attribute;
+        });
+    typeConverter.addTypeAttributeConversion(
+        [&typeConverter](ShapedType type,
+                         DenseFPElementsAttr attr) -> Attribute {
+          return convertDenseElementsAttr<Kind, DenseFPElementsAttr, FloatType>(
+              type, attr, typeConverter,
+              [](FloatType newElementType, const APFloat &value) {
+                APFloat converted(value);
+                bool losesInfo = false;
```

Comment:
nit: This looks quite similar to `tryconvertScalarAttribute` above, curious if we can factor this into a common function?